src/StaticLS/HIE/File.hs

```diff
@@ -0,0 +1,176 @@
+import qualified Data.Map as Map
+import qualified GHC
+import GHC.Data.Maybe (firstJusts)
+import qualified GHC.Iface.Ext.Binary as GHC
+import qualified GHC.Iface.Ext.Types as GHC
+import qualified GHC.Types.Name.Cache as GHC
+import qualified HieDb
+import qualified Language.LSP.Types as LSP
+import StaticLS.HIE.File.Except
+import StaticLS.Maybe (flatMaybeT)
+import StaticLS.StaticEnv
+import qualified System.Directory as Dir
+import System.FilePath ((-<.>), (</>))
+
+type SrcFilePath = FilePath
+type HieFilePath = FilePath
+
+-- | Retrieve a hie info from a lsp text document identifier
+getHieFileFromTdi :: (HasStaticEnv m, MonadIO m) => LSP.TextDocumentIdentifier -> ExceptT HieFileTdiException m GHC.HieFile
+getHieFileFromTdi tdi = do
+    srcFilePath <- maybe (throwE HieTdiSrcNotFoundException) pure $ LSP.uriToFilePath tdi._uri
+    hieFilePath <- maybeT (throwE HieTdiHieNotFoundException) pure $ srcFilePathToHieFilePath srcFilePath
+    withExceptT HieTdiReadException $ getHieFile hieFilePath
+
+-- | Retrieve an hie file from a hie filepath
+getHieFile :: (HasStaticEnv m, MonadIO m) => HieFilePath -> ExceptT HieFileReadException m GHC.HieFile
+getHieFile hieFilePath = do
+    staticEnv <- getStaticEnv
+    ExceptT . liftIO $
+        (pure . Right . GHC.hie_file_result <=< GHC.readHieFile staticEnv.nameCache) hieFilePath
+            `catch` (pure . Left . HieFileReadException)
+
+-- | Retrieve an hie file from a module name
+modToHieFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> MaybeT m GHC.HieFile
+modToHieFile = exceptToMaybeT . getHieFile <=< modToHieFilePath
+
+-- | Retrieve a src file from a module name
+modToSrcFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> MaybeT m SrcFilePath
+modToSrcFile = hieFilePathToSrcFilePath <=< modToHieFilePath
+
+{- | Fetch a src file from an hie file, checking hiedb but falling back on a file manipulation method
+if not indexed
+-}
+srcFilePathToHieFilePath :: (HasStaticEnv m, MonadIO m) => SrcFilePath -> MaybeT m HieFilePath
+srcFilePathToHieFilePath srcPath = do
+    t1 <- runMaybeT $ srcFilePathToHieFilePathHieDb srcPath
+    t2 <- runMaybeT $ srcFilePathToHieFilePathFromFile srcPath
+    MaybeT $ pure $ firstJusts [t1, t2]
+
+-- | Fetch an hie file from a src file
+hieFilePathToSrcFilePath :: (HasStaticEnv m, MonadIO m) => HieFilePath -> MaybeT m SrcFilePath
+hieFilePathToSrcFilePath = hieFilePathToSrcFilePathFromFile
+
+-----------------------------------------------------------------------------------
+-- HieDb Method of file lookups - requires hiedb to be indexed using --src-base-dirs from 0.4.4.0
+-----------------------------------------------------------------------------------
+
+srcFilePathToHieFilePathHieDb :: (HasStaticEnv m, MonadIO m) => SrcFilePath -> MaybeT m HieFilePath
+srcFilePathToHieFilePathHieDb srcPath = do
+    absSrcPath <- MaybeT . fmap Just . liftIO $ Dir.makeAbsolute srcPath
+    hieModRow <- flatMaybeT $ runHieDbMaybeT $ \hieDb -> do
```

Comment:
A little trick is that you can use the `MonadFail` instance for `MaybeT` to do a partial pattern match on bind and it's safe for `MaybeT` (it translates to `Nothing` if the pattern match fails):

```suggestion
    Just hieModRow <- runHieDbMaybeT $ \hieDb -> do
```